docs/codebase-best-practices.md

```diff
@@ -0,0 +1,65 @@
+# Codebase Best Practices
+
+## General JavaScript
+
+In most cases, our [linter](how-to-setup-freecodecamp-locally?id=follow-these-steps-to-get-your-development-environment-ready) will warn of any formatting which goes against this codebase's preferred practice.
+
+## Specific TypeScript
+
+### Migrating a JavaScript File to TypeScript
+
+#### Retaining Git File History
+
+Sometimes changing the file from `<filename>.js` -> `<filename>.tsx?` causes the original file to be deleted, and a new one created, and other times the filename just changes - in terms of Git. Ideally, we want the file history to be preserved.
+
+The best bet at achieving this is to:
+
+1. Rename the file
+2. Commit with the flag `--no-verify` to prevent Husky from complaining about the lint errors
+3. Refactor to TypeScript for migration, in a separate commit
+
+> [!NOTE]
+> Editors like VSCode are still likely to show you the file has been deleted and a new one created. If you use the CLI to `git add .`, then VSCode will show the file as renamed in stage
+
+### Naming Conventions
+
+#### Types
+
+React Component Props
+
+```typescript
+type MyComponentProps = {};
+const MyComponent = (props: MyComponentProps) => {};
+```
+
+Default
+
+```typescript
+type MyObjectType = {};
+const myObject: MyObjectType = {};
+```
+
+#### Interfaces
+
+React Component Props
+
+```typescript
+interface IMyComponentProps {}
+const MyComponent = (props: IMyComponentProps) => {};
```

Comment:
```suggestion
interface MyComponentProps {}
const MyComponent = (props: MyComponentProps) => {};
```
This seems to not be the convention: https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#names

The rationale is basically that, aside from `extends`, interfaces and types are basically the same.  And, if someone tries to `extend` a type TS will tell them, so we don't have to.